### PR TITLE
Add booster tests to CI

### PR DESCRIPTION
**Pull request by goodlyrottenapple**
The booster backend is now getting to the state where we can start testing against it for foundry proofs. This PR pulls in the booster from the private cache and executes the foundry tests. Running with booster can be enabled for kevm foundry by first compiling with `kevm foundry-kompile ... --with-llvm-library` and then `kevm foundry-prove ... --use-booster`

### CODE REVIEW

File: kevm-pyk/src/kevm_pyk/foundry.py [hunk 1]
```diff
@@ -816,11 +837,6 @@ def _method_to_apr_proof(
     reinit: bool = False,
     simplify_init: bool = True,
     bmc_depth: int | None = None,
-    kore_rpc_command: str | Iterable[str] = ('kore-rpc',),
-    smt_timeout: int | None = None,
-    smt_retry_limit: int | None = None,
-    bug_report: BugReport | None = None,
-    trace_rewrites: bool = False,
```
Comment (goodlyrottenapple) [inline kevm-pyk/src/kevm_pyk/foundry.py:823]: these were unused

File: .github/workflows/test-pr.yml [hunk 2]
```diff
@@ -169,6 +169,52 @@ jobs:
...
+      - name: 'Build Foundry'
+        run: nix develop --extra-experimental-features 'nix-command flakes' --command make build-foundry -j2
+      - name: 'Foundry Kompile'
+        run: nix develop --extra-experimental-features 'nix-command flakes' --command make test-foundry-kompile -j2 KEVM_OPTS=--with-llvm-library
+      - name: 'Foundry Prove'
+        run: |
+          BOOSTER_BIN="$(pwd)/booster/result/bin/booster"
+          nix develop --extra-experimental-features 'nix-command flakes' --command make test-foundry-prove -j2 FOUNDRY_PAR=12 KPROVE_OPTS="--use-booster $BOOSTER_BIN"
+      - name: 'Cleanup'
+        run: nix develop --extra-experimental-features 'nix-command flakes' --command make clean
```
Comment (goodlyrottenapple) [inline .github/workflows/test-pr.yml:216]: not sure if this step is necessary? we shouldn't be persisting the `.build` folder afaik, given that it lives in the checked out repo folder.

Comment (ehildenb) [reply .github/workflows/test-pr.yml:214]: It shouldn't be necessary I don't think.

File: .github/workflows/test-pr.yml [hunk 3]
```diff
@@ -186,6 +209,8 @@ jobs:
           install_url: https://releases.nixos.org/nix/nix-2.13.3/install
           extra_nix_config: |
             access-tokens = github.com=${{ secrets.GITHUB_TOKEN }}
+            substituters = http://cache.nixos.org https://cache.iog.io
+            trusted-public-keys = cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY= hydra.iohk.io:f/Ea+s+dFdN+3Y/G+FDgSq+a5NEWhJGzdjvKNGv0/EQ=
       - name: 'Install Cachix'
```
Comment (ehildenb) [inline .github/workflows/test-pr.yml:214]: Is this change necessary? Why so?

Comment (goodlyrottenapple) [reply .github/workflows/test-pr.yml:214]: we are implicitly building the haskell/booster backend which relies on the iog cache. best keep it in on the public runner. our self hosted runners (should) have the same cache set up

File: kevm-pyk/src/kevm_pyk/__main__.py [hunk 4]
```diff
@@ -922,6 +924,13 @@ def parse(s: str) -> list[T]:
         type=int,
         help='Max depth of loop unrolling during bounded model checking',
     )
+    foundry_prove_args.add_argument(
+        '--use-booster',
+        dest='use_booster',
```
Comment (tothtamas28) [inline kevm-pyk/src/kevm_pyk/__main__.py:929]: I think that's the default destination for this flag.

File: kevm-pyk/src/tests/integration/test_foundry_prove.py [hunk 5]
```diff
@@ -35,12 +35,15 @@ def foundry_root(tmp_path_factory: TempPathFactory) -> Path:
...
+        llvm_library=use_booster,
     )
 
     return foundry_root
 
 
-def test_foundry_kompile(foundry_root: Path, update_expected_output: bool) -> None:
+def test_foundry_kompile(foundry_root: Path, update_expected_output: bool, use_booster: bool) -> None:
+    if use_booster:
+        return
```
Comment (tothtamas28) [inline kevm-pyk/src/tests/integration/test_foundry_prove.py:46]: `use_booster` should only modify the RPC server started by the implementation. If you need more control on which checks are performed, consider creating a dedicated test module (e.g. `test_booster.py`).

File: kevm-pyk/src/tests/integration/test_foundry_prove.py [hunk 6]
```diff
@@ -88,12 +91,13 @@ def test_foundry_prove(test_id: str, foundry_root: Path, update_expected_output:
...
         smt_timeout=125,
         smt_retry_limit=4,
+        use_booster=use_booster,
     )
 
     # Then
     assert_pass(test_id, prove_res)
 
-    if test_id not in SHOW_TESTS:
+    if test_id not in SHOW_TESTS or use_booster:
```
Comment (tothtamas28) [inline kevm-pyk/src/tests/integration/test_foundry_prove.py:100]: Same here.

File: kevm-pyk/src/tests/integration/test_foundry_prove.py [hunk 7]
```diff
@@ -116,20 +120,21 @@ def test_foundry_prove(test_id: str, foundry_root: Path, update_expected_output:
...
         smt_timeout=125,
         smt_retry_limit=4,
+        use_booster=use_booster,
     )
 
     # Then
     assert_fail(test_id, prove_res)
 
-    if test_id not in SHOW_TESTS:
+    if test_id not in SHOW_TESTS or use_booster:
```
Comment (tothtamas28) [inline kevm-pyk/src/tests/integration/test_foundry_prove.py:137]: And here.

File: kevm-pyk/src/kevm_pyk/foundry.py [hunk 8]
```diff
@@ -425,6 +426,31 @@ def foundry_prove(
...
+            kore_rpc_command = ('kore-rpc-booster', '--llvm-backend-library', str(llvm_dylib))
+        else:
+            llvm_dylib = foundry_llvm_dir / 'interpreter.dylib'
+            if llvm_dylib.exists():
+                kore_rpc_command = ('kore-rpc-booster', '--llvm-backend-library', str(llvm_dylib))
+            else:
+                raise ValueError(
+                    f"Could not find the LLVM dynamic library in {foundry_llvm_dir}. Please re-run foundry-kompile with the '--with-llvm-library' flag"
+                )
+
```
Comment (tothtamas28) [inline kevm-pyk/src/kevm_pyk/foundry.py:453]: Consider extracting a method / (cached) property on `Foundry`:

```python
def llvm_dylib(self) -> Path
```

File: kevm-pyk/src/tests/integration/test_foundry_prove.py (hunk 5 above)
Comment (goodlyrottenapple) [reply kevm-pyk/src/tests/integration/test_foundry_prove.py:46]: these are all necessary because the golden files will not exactly correspond to the output of the booster. when use_booster is enable all we are doing is testing that the tests pass.

Comment (tothtamas28) [reply kevm-pyk/src/tests/integration/test_foundry_prove.py:46]: I see, makes sense.
